Add /recent endpoint with recent blocks and forks

JamesPiechota opened the pull request:
The `blocks` field stores the last 18 blocks that have been received including their timestamp. The timestamps of the most recent 5 are hidden and when a miner mines a block it uses the timestamp of the first time that same block is shared  to it. These 2 things are done so that we don't leak information about which blocks a miner has mined.

The `forks` field lists all recently received forks - the forks are persisted to disk as well. A fork is identified by:
`block_ids`: a llist of block ids for all abandoned blocks
`id`: a hash of the `block_ids` that serves as the fork's id
`height`: the height of the first abandoned block
`timestamp`: the timestamp (in seconds) of when the fork was detected

JamesPiechota committed:
Have ar_http_iface_client::get_info() use a map

Previously it manually built a JSONStruct/PropList and used that
intermediate format to store/access individual info items. However
we'll soon be adding recent and forks keys which will have more
complex structures. Manually building and searching a proplist will
get hairy. This commit moves to using an erlang map as the
intermediate format and relying on the jiffy encode/decode funcitons
for converting to/from json (rather than doing it by hand).

JamesPiechota committed:
Add recent blocks to /info

JamesPiechota committed:
Simplify /info and /recent serialization

JamesPiechota committed:
Add a `forks` field to /recent that returns recently detected forks

ldmberman commented on apps/arweave/src/ar_http_iface_middleware.erl:
```diff
@@ -2358,6 +2321,7 @@ handle_block_announcement(#block_announcement{ indep_hash = H, previous_block =
 		solution_hash = SolutionH }, Req) ->
 	case ar_ignore_registry:member(H) of
 		true ->
+			ar_block_cache:update_timestamp(block_cache, H, erlang:timestamp()),
```
Why do we update the timestamp here? We've seen the block before so shouldn't we have the relevant timestamp assigned already?

JamesPiechota replied:
Mostly to handle the fact that we only want to record the timestamp the first time the block is received from a peer so that we don't record the timestamp when the block was first mined. So in the event that this node mined a block it'll be added to the ignore registry even though the received timestamp hasn't been set. Instead we always check the timestamp whenever we receive an announcement or post, and update it if it hasn't already been set.

JamesPiechota replied:
since we do an ETS lookup by a specific key, I'm assuming the performance impact of the extra check whenever we handle a post or announcement is minimal . But perhaps that's a bad assumption?

ldmberman replied:
Are there any other places where we call update_timestamp except the two I see here under ar_ignore_registry:member(H) == true? Does it mean we only set the timestamp the second time we receive the block?

JamesPiechota replied:
We also set the timestamp the first time a block is pre-validated.

The value gets set in the Block here: https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_block_pre_validator.erl#L49

And the block gets cached when the `{block, {new` event is handled, sent here: https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_block_pre_validator.erl#L746

JamesPiechota replied:
ugh, I suppose I also need to handle the first time that a node fetches a block it had previously mined: https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_poller.erl#L146

JamesPiechota replied:
Basically: it's the handling of self-mined blocks that causes all the trouble since we *don't* want the timestamp to be set to the momen the block was mined since that will make it possible to determine which nodes mined which blocks. And near as I can tell there are 3 places were we receive a block from a peer and need to check the timestamp:
- block announcement
- POST block
- GET block via polling
-

JamesPiechota replied:
Atually never mind, I don't think polling is relevant for this case. If the node has already mined the block it won't poll it from a peer, and if it hasn't mined the block then the timestamp will be handled via pre-validation when it polls it.

JamesPiechota replied:
So yeah, as far as I can tell:

Previous behavior:
1. receive_timestamp is *not* set when a node mines a block
2. receive_timestamp is set when a newly received block passes pre-validation

Desired behavior:
1. always set receive_timestamp the first time a node receives a block from a peer
2. To do this we need to handle the case where a block exists in a node's block_cache because it was the one who mined it. In this case we need to update the receive_timestamp when it receives the block either via `POST /block` or `POST /block_announcement` even though that block make already exist in the ar_ignore_registry

Let me know if you I've missed something or you think it should be implemented differently

ldmberman replied:
Ah, yes, I remember now, thank you for the detailed explanation!

ldmberman commented on apps/arweave/src/ar_info.erl:
```diff
@@ -0,0 +1,94 @@
+            }]
+        end,
+        [],
+        lists:sublist(ar_block_index:get_list(CurrentHeight), ?CHECKPOINT_DEPTH)
```
ar_block_index:get_list is very slow as it fetches the entire block index - we may use e.g. [ar_node:get_block_anchors](https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_node.erl#L183C1-L183C18) to quickly fetch the recent block hashes.

JamesPiechota replied:
Updated: https://github.com/ArweaveTeam/arweave/pull/581/commits/4764260fed3c794b8e5c17574edb43476ae51717

ldmberman commented on apps/arweave/src/ar_chain_stats.erl:
```diff
@@ -0,0 +1,176 @@
+    process_flag(trap_exit, true),
+    Columns = [
+        {"default", []},
+        {"forks", []}
```
Do we need to use a custom RocksDB column? The feature has a bug where the data in the non-default columns would not survive the first non-graceful shutdown (some problems with WAL population, perhaps).

JamesPiechota replied:
Oh. Ok. So I should create a separate rocksdb file for each column? My plan is to add a few more columns in the ar_chai_stats module to track some other data needed for separate bugs/requests

ldmberman replied:
I would do one db per table, like we have it e.g., in [ar_storage](https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_storage.erl#L1030), do you see major downsides to this approach?

JamesPiechota replied:
Nope! I was just trying to follow conventions and wasn't sure whether ar_data_sync or ar_storage was the "right" way. I'll do one file per table!

JamesPiechota replied:
Updated: https://github.com/ArweaveTeam/arweave/pull/581/commits/b63151dc1abd41661302bb5fce2f810f7a5cd7dc

ldmberman commented on apps/arweave/src/ar_chain_stats.erl:
```diff
@@ -0,0 +1,176 @@
+%% @doc Returns all forks that have been logged since the given start time
+%% (system time in seconds)
+get_forks(StartTime) ->
+    gen_server:call(?MODULE, {get_forks, StartTime}).
```
I'd handle the timeout gracefully here, I believe, we'll hit it every once in a while.

JamesPiechota replied:
Updated: https://github.com/ArweaveTeam/arweave/pull/581/commits/b63151dc1abd41661302bb5fce2f810f7a5cd7dc

ldmberman commented on apps/arweave/src/ar_info.erl:
```diff
@@ -0,0 +1,94 @@
+            }]
+        end,
+        [],
+        ar_chain_stats:get_forks(0)
```
Do we fetch all forks here, not just a recent subset?

JamesPiechota replied:
Now we do! Update to fetch just a recent subset: https://github.com/ArweaveTeam/arweave/pull/581/commits/b63151dc1abd41661302bb5fce2f810f7a5cd7dc

JamesPiechota replied:
I may update this logic again when I switch to using get_block_anchors - but I'll do that in a separate commit

ldmberman commented on apps/arweave/src/ar_http_iface_client.erl:
```diff
@@ -1086,8 +1086,7 @@ get_info(Peer, Type) ->
 		Info ->
-			{Type, X} = lists:keyfind(Type, 1, Info),
-			X
+			maps:get(atom_to_binary(Type), Info)
```
The previous implementation handled the case when an expected key would not be sent (plus the absence of the release field treated as release=0), here we will have badarg, if I see it right.

JamesPiechota replied:
what about if I change to `maps:get(atom_to_binary(Type), Info, info_unavailable)`?

I believe that's what happened previously if the expected key was not sent.

In that case the only situation that isn't handled is the special handling for a missing `release`. Which perhaps I can handle in `ar_peers.erl` (the only place where `get_info(Peer, release)` is called?

My motivation in changing this to a map was to cut down on some of the boilerplate that we had where we converted JSON -> JSonStruct -> Map -> value.

Let me build a commit to show you the proposed fix.

JamesPiechota replied:
Proposed fix: https://github.com/ArweaveTeam/arweave/pull/581/commits/319984a85aba444a1492271f22f53ba8010f8898

ldmberman replied:
Makes sense, looks good, thank you!

ldmberman commented on apps/arweave/src/ar_block_cache.erl:
```diff
@@ -456,6 +456,27 @@ get_siblings(Tab, B) ->
+		[{_, {B, Status, Timestamp, Children}}] ->
+			case B#block.receive_timestamp of
+				undefined ->
+					insert(Tab, {
```
It's not thread-safe to update this table outside the ar_node_worker process.

JamesPiechota replied:
ah, oof. Okay I'll rework it

JamesPiechota replied:
I moved the update into ar_node_worker: https://github.com/ArweaveTeam/arweave/pull/581/commits/e6b2dd18b55d769aa62d2bfb187db946c5d3b48c

JamesPiechota committed:
Fix a number of bugs identified in the PR review
Better error handling
Fix potential race condition in ar_block_cache update
Limit number of forks returned
Fix bugs in the ordering of blocks and forks
Add more tests

ldmberman commented on apps/arweave/src/ar_chain_stats.erl:
```diff
@@ -0,0 +1,173 @@
+%%%===================================================================
+log_fork(Orphans, ForkRootB, ForkTime) ->
+    Fork = create_fork(Orphans, ForkRootB, ForkTime),
+    ar_kv:put(forks_db, integer_to_binary(ForkTime), term_to_binary(Fork)),
```
integer_to_binary(2) > integer_to_binary(11); I think, we need << ForkTime:64 >> here

JamesPiechota replied:
Yikes good catch. I added a test and fixex: https://github.com/ArweaveTeam/arweave/pull/581/commits/0542b4e1254d1d50b751c3b1589fb995e9b7f3a8

ldmberman commented on apps/arweave/src/ar_info.erl:
```diff
@@ -0,0 +1,99 @@
+        %%   "timestamp": <timestamp_of_when_fork_was_abandoned>
+        %%   "blocks": [<block_id>, <block_id>, ...]
+        %% }
+        <<"forks">> => get_recent_forks()
```
The {error, _} response will still crash the handler process

JamesPiechota replied:
Got it. I haven't written the tests for that bit yet. I'll do that next and then fix.

JamesPiechota replied:
I think I got it this time? https://github.com/ArweaveTeam/arweave/pull/581/commits/0542b4e1254d1d50b751c3b1589fb995e9b7f3a8

ldmberman commented on apps/arweave/src/ar_info.erl:
```diff
@@ -0,0 +1,100 @@
+                    #fork{ 
+                        id = ID, height = Height, timestamp = Timestamp, 
+                        block_ids = BlockIDs} = Fork,
+                    Acc ++ [#{
```
20 times is not a lot, but I would avoid some unnecessary reductions here (e.g., might matter slightly under load) and prepend `[#{...} | Acc]`, with lists:reverse if we want this ordering.

JamesPiechota replied:
Updated! https://github.com/ArweaveTeam/arweave/pull/581/commits/9e18942770c521b4dcb50e31cc00641a2cdea790

ldmberman commented on apps/arweave/src/ar_node.erl:
```diff
@@ -181,8 +181,12 @@ get_current_usd_to_ar_rate() ->
 %% the hashes of the recent blocks that can be used in transactions as anchors.
 %% @end
 get_block_anchors() ->
-	[{block_anchors, BlockAnchors}] = ets:lookup(node_state, block_anchors),
```
Could we return an atom here (e.g., not_joined) to avoid bugs?

ldmberman replied:
Hm, I think, it makes sense to return 403 not_joined from the new endpoint until the node has been initialized, like in most other endpoints?

JamesPiechota replied:
Updated! https://github.com/ArweaveTeam/arweave/pull/581/commits/9e18942770c521b4dcb50e31cc00641a2cdea790

JamesPiechota commented:
@ldmberman Added fix to your comment about recent blocks ordering (I can't find the comment anymore though...). https://github.com/ArweaveTeam/arweave/pull/581/commits/1a2d357299ca23b2a4753f272eec9589e0a16581

6th time's the charm ;)

ldmberman reviewed (approved):
LGTM, thank you!
